keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,366 @@
+
+| Impact | Description|
+| ------- | ------------ |
+| + | Reliability - for various reasons kube-API might not be available but this doesn’t mean that local workloads are not accessible and on node system workloads should have the most recent data about pod's readiness even when kube-API is unreachable. |
+| + | Scalability - Reduce the load on kube-API be reducing the number of watchers and using kubelet to fetch local pods data. Fetching only pods limited to one node is costly operation for kube-API. |
+| + | Safety - Read-only API will not add security risks. |
+| + | Reduce resource consumption by workload. Using Kube-API we can fetch objects like PodSpec or PodStatus, for some on-node workload this might be unnecessary, with this API workload can reduce the resource consumption. |
+| - | This API will add load to kubelet (mitigation: API will be rate limited) |
+| - | Kubelet does not support RBAC authorization for gRPC. (mitigation: This API is design to be accessible for all workloads running on the node without the authorization. The unix socket will be used for the connection and all exposed data will be carefully reviewed.) |
+
+### Goals
+
+The goal of this API is to expose Pod readiness information directly from the source - kubelet, independent on Control Plane availability. This would remove the need for node-local components to request this node-local information from the Kubernetes API Server. Kubelet already has a podresources endpoint ([2403-pod-resources-allocatable-resources](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2403-pod-resources-allocatable-resources)) which returns information about Pod’s containers and Devices. This API does not contain information about pod readiness status. Kubelet is responsible for computing pod status and stores it in a local cache. We want to create a new API that will expose pod conditions that are computed by kubelet and return the most recent data even when kube-API is not reachable.
+
+### Non-Goals
+
+### User Stories
+* Some on node system workloads want to reduce Control Plane dependency and introduce locality for pod’s readiness due to reliability and scalability issues.
+* For building custom monitoring tools that show information about pod’s readiness. When it is ready to serve the traffic.
+* Some on node system workloads interested in pods readiness want to reduce resource consumption.
+
+## Proposal
+
+We are proposing to create a new kubelet API that will return pod's readiness information.
+* The API will return data about both Static and Regular Pods.
+* The API will not return partial data. If kubelet does not know actual information about workloads then appropriate error code will be returned.
+* API should return the most recent information about pods computed by the kubelet even when those data were not reported or accepted by kube-API.
```

Comment:
The fundamental purpose of this KEP is to ensure that there is a path to get Pod readiness information without a dependency on the Kubernetes control plane. One of the great parts about Kubernetes is that it can coast reasonably well if the control plane is unavailable. One problem is that if Pod readiness ever changes in that state, networking components are going to be completely unaware of that because it has a control plane dependency. Removing that dependency as proposed here may improve both the reliability and performance of the system.